### PR TITLE
WIP: early_stopping_rounds added

### PR DESCRIPTION
**Pull request by kryptonite0**
Implemented basic logic for early stopping. 

If the model parameter `early_stopping_rounds` is set, the training stops after such number of iterations pass without loss improvement. 
If `early_stopping_rounds` is not set, the previously implemented stopping logic is used.

Not sure what's the best way to return the model from the best iteration. What's the current implementation doing in this regard? I mean, after `.fit()`, what iteration does the model use to `.predict()`?

### CODE REVIEW

**Comment by alejandroschuler** [comment]
Hey @kryptonite0 I [took your code and modified it](https://github.com/stanfordmlgroup/ngboost/commit/32e9736323b282f524274b7d88b05f0e84f987c8) a) to make it a little more efficient and b) so that `early_stopping_rounds` is an argument to `fit` instead of to the constructor, to be in line with how it's done in xgboost and keep the API cleaner. I also removed the existing stopping logic altogether, which I think is redundant with what you added.

The problem of returning the predictions at that iteration is not yet addressed- I'll open an issue specifically for that.

**Comment by civilinformer** [comment]
I am not sure why xgboost does it that way. For instance, Lightgbm using the Sklearn API, sets early stopping in the constructor. In fact if you want to follow the sklearn API, all parameters should be set in the constructor. If you want to use sklearn's cross validation, it makes it less clean to pass parameters via fit.

**Comment by alejandroschuler** [comment]
Ah... well I should have done a little more research :) If you open a pull request to stick it back in the constructor I will readily approve it- just remember to also reflect the change in api.py.

**Comment by kmedved** [comment]
Is there currently any way to access to the number of iterations used for early stopping? For instance, in LightGBM, you can access the number of rounds from the best iteration via a  `.best_iteration_` call in the sklearn API.

This is very helpful for preventing overfitting, so you can see the number of rounds at which the model began overfitting on your validation data, so you can later pass that best_iteration count as the `n_estimators` count for your test data. I don't currently see a way to access this number however.

Relatedly, with respect to the discussion above, LightGBM follows the xgboost approach of passing this parameter [into the .fit() call](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor.fit), so I believe the current approach is consistent with LightGBM. 

That said, I agree with @civilinformer that it makes more sense to be in the constructor, so it can be used with sklearn's cross-validation. Right now, I don't see a good way to pass `early_stopping_rounds` into [`model_selection.cross_validate`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate), so you need to write a custom cross validation function to take advantage of that functionality.

Further confusing matters, sklearn is rolling out their own version of LightGBM called [`HistGradientBoostingRegressor`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.HistGradientBoostingRegressor.html), and **that** puts the `n_iter_no_change` parameter into the constructor (where I think it makes more sense). And of course, the non-histogram `GradientBoostingRegressor` likewise puts the `n_iter_no_change` parameter into the constructor as well.

[Catboost also follows the sklearn](https://catboost.ai/docs/concepts/python-reference_catboostregressor.html) approach of putting this in the constructor. It seems like xgboost and LightGBM are the outliers among the major boosting libraries in this respect, although I don't know why.

**Comment by alejandroschuler** [comment]
Yes, that can be accessed through the fit model's `self.best_val_loss_itr` instance attribute. Let me look into the CV issue and I'll get back to you on that shortly. Will probably move it into the constructor as you suggest.

**Comment by alejandroschuler** [comment]
At first glance, though, I don't think early stopping will play nice with sklearn CV, which is kind of dumb. Early stopping requires that the validation data be passed to `fit` which sklearn doesn't support iirc.

**Comment by kmedved** [comment]
Frustrating! Given how common early stopping schemes are these days, that would seem to be a natural functionality to support. In that case, I'm not really sure what the benefit is in moving the attribute into the constructor, although I do think it's more sklearn-like in that sense (although I am far from an expert).

Thanks @alejandroschuler

**Comment by civilinformer** [comment]
In the context of a complicated CV it would make sense for `fit`, given that the number of stopping rounds is set, to set aside a fraction of the regular training data, also determined by a parameter in the constructor, to use as stopping guidance.

**Comment by alejandroschuler** [comment]
You could do that. If used with sklearn CV it would basically do an inner validation on each training fold and then "test" on each validation fold. I guess that's ok though.

**Comment by civilinformer** [comment]
Anybody could decorate the class to achieve that I guess, so maybe early stopping is not the highest priority? While I am commenting just interject that any speed improvements would probably be more helpful, but I understand that they are not in the works...
